### PR TITLE
Revert LineIndex optimizations to fix #5656

### LINKED ISSUE #5656: False positive syntax errors
I just updated rust-analyzer this morning, and it's no long working for me.
If I first open VS Code, rust-analyzer doesn't give any errors on my codebase (nor does `cargo check`). However, as soon as I begin typing (or even saving), I get persistent false positive syntax errors that completely break syntax highlighting and suggestions.  `cargo clean` doesn't resolve the problem.

Please let me know if there is any additional information that might clear this up. I can try re-installing rust-analyzer fresh, but figured I would post an issue in case this isn't affecting just me.

For instance: top image is right after I have opened VS code, and bottom is as soon as I have started typing. Deleting the line that I was typing does not make the errors go away, nor does saving. The only way to refresh the syntax errors is to restart VS Code.

Version numbers of relevant software:
```
rustc: 1.45.0 5c1f21c3b 2020-07-13) [stable]
rust-analyzer: 0.2.264
```

```
VS Code: 1.47.3 (system setup)
Commit: 91899dcef7b8110878ea59626991a18c8a6a1b3e
Date: 2020-07-23T13:12:49.994Z
Electron: 7.3.2
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Windows_NT x64 10.0.18362
```

### PR DESCRIPTION
**Pull request by lazear**
This PR reverts the changes from https://github.com/rust-analyzer/rust-analyzer/pull/5532, which was causing issues as described in
https://github.com/rust-analyzer/rust-analyzer/issues/5656

### CODE REVIEW

File: crates/rust-analyzer/src/main_loop.rs [hunk 1]
```diff
@@ -424,13 +423,11 @@ impl GlobalState {
                 if let Ok(path) = from_proto::vfs_path(&params.text_document.uri) {
                     let DidChangeTextDocumentParams { text_document, content_changes } = params;
                     let vfs = &mut this.vfs.write().0;
-                    let world = this.snapshot();
                     let file_id = vfs.file_id(&path).unwrap();
 
                     // let file_id = vfs.file_id(&path).unwrap();
```
Comment (lnicola) [inline crates/rust-analyzer/src/main_loop.rs:428]: Can you also remove this commented line (or use git revert)?

Comment (lazear) [reply crates/rust-analyzer/src/main_loop.rs:428]: Yes I can remove the commented line

**Comment by lazear** [comment]
I am inclined to agree that it might be something deeper, especially if others are still having the issue

File: crates/rust-analyzer/src/main_loop.rs (hunk 1 above)
Comment (lazear) [reply crates/rust-analyzer/src/main_loop.rs:428]: @lnicola is there a specific way to use `git revert` for this? I haven't used it before.

**Comment by lnicola** [comment]
We could also add an assert that compares the old index (the parameter) with a newly-constructed one. Maybe that would provide a hint.

**Comment by lnicola** [comment]
For context, the result of the line_index call is computed the first time for each LSP request, then cached. That's why there's no set_line_index method.

This optimization tried to reuse the existing index when applying an edit. Since most edits affect a single range (unless you're e.g. using multiple cursors), that PR avoided recomputing the index.

I wonder if we're getting a wrong file id (I had to shuffle around some code, look at the original PR). It's interesting how you don't see this in the RA code.

**Comment by lnicola** [comment]
git reset --hard master
git revert 83a87fcd11298d8de32801ac6496ddf8b4e9c0e1
git push -f

**Comment by lazear** [comment]
@lnicola 
> git reset --hard master
> git revert [83a87fc](https://github.com/rust-analyzer/rust-analyzer/commit/83a87fcd11298d8de32801ac6496ddf8b4e9c0e1)
> git push -f

Done.
Had I known this was possible, I would've just done that to start with. TIL 😃.

**Comment by lnicola** [comment]
Unfortunately, I don't have commit rights.

Is your code public, in case I manage to reproduce the issue and look into the line index differences?

**Comment by lazear** [comment]
Sure thing. Issues were occurring on any of the files in the `sml-core` crate here:
https://github.com/SomewhatML/sml-compiler/blob/master/crates/sml-core/src/match_compile.rs

**Comment by matklad** [comment]
I guess the issue here is that we might call this *before* we call `process_changes`? Ie, we have new data in VFS but old data in salsa?

I *think* there might be was some reason why I removed this caching optimization in the first place :D

**Comment by CAD97** [comment]
I ran into a server panic due to document state getting out of sync ([zulip](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Fwg-rls-2.2E0/topic/lsp.20got.20out.20of.20sync)) that appears was fixed by this patch, probably due to the same issue. It can cause some bizarre behavior with fmt-on-save replacing the document with completely different contents sometimes 😅

Perhaps my repo can serve as a somewhat smaller repro case, and I have a consistent one.

Repo: https://github.com/cad97/cfg-pub
Repro example: https://imgur.com/72UJjjK

Steps:

- Enable fmt-on-save
- Edit [./cfg-pub/examples/example.rs](https://github.com/CAD97/cfg-pub/blob/main/cfg-pub/examples/example.rs) to add a blank fn definition below the `main`
- Load a bad version of r-a
- Make an edit and save
- Observe strange formatting applied by save
- Fix formatting weirdness
- Observe the IDE fighting to ruin it more (or just panicking and dying)

**Pull request merged**